File: bindings/python/quad_key/tests/test_geo_coordinate2d.py
"""
Unit Tests for geo_coordinate2d
"""

from __future__ import print_function

import unittest

from quad_key.quad_key import geo_coordinate2d
from quad_key.quad_key import longitude
from quad_key.quad_key import latitude


class TestGeoCoordinate2d(unittest.TestCase):
    """Class for running unit tests."""

    def test_default_ctor(self):
        """Test default ctor."""
        empty = geo_coordinate2d()
        self.assertEquals(longitude(0.0), empty.get_longitude())
        self.assertEquals(latitude(0.0), empty.get_latitude())

    def test_ctor(self):
        """Test ctor."""
        coord = geo_coordinate2d(longitude(0.0), latitude(0.0))
        self.assertEquals(longitude(0.0), coord.get_longitude())
        self.assertEquals(latitude(0.0), coord.get_latitude())

        coord = geo_coordinate2d(longitude(-180.001), latitude(-95.0))
        self.assertEquals(longitude(-180.0), coord.get_longitude())
        self.assertEquals(latitude(-90.0), coord.get_latitude())

        coord = geo_coordinate2d(longitude(-180.001), latitude(95.0))
        self.assertEquals(longitude(-180.0), coord.get_longitude())
        self.assertEquals(latitude(90.0), coord.get_latitude())

        coord = geo_coordinate2d(longitude(180.001), latitude(95.0))
        self.assertEquals(longitude(180.0), coord.get_longitude())
        self.assertEquals(latitude(90.0), coord.get_latitude())

        coord = geo_coordinate2d(longitude(180.001), latitude(-95.0))
        self.assertEquals(longitude(180.0), coord.get_longitude())
        self.assertEquals(latitude(-90.0), coord.get_latitude())

    def test_near(self):
        """Test to see if one coord is near another."""
        lon = longitude(0.001)
        self.assertTrue(lon.near(longitude(0.0009), 1e-3))

        lon = longitude(0.001)
        self.assertFalse(lon.near(longitude(0.0009), 1e-4))

        lat = latitude(0.001)
        self.assertTrue(lat.near(latitude(0.0009), 1e-3))

        lat = latitude(0.001)
        self.assertFalse(lat.near(latitude(0.0009), 1e-4))

        coord = geo_coordinate2d(longitude(0.001), latitude(0.001))
        self.assertTrue(coord.near(
            geo_coordinate2d(longitude(0.0009), latitude(0.0009)), 1e-3))

        coord = geo_coordinate2d(longitude(0.001), latitude(0.001))
        self.assertFalse(coord.near(
            geo_coordinate2d(longitude(0.001), latitude(0.0009)), 1e-4))

        coord = geo_coordinate2d(longitude(0.001), latitude(0.001))
        self.assertFalse(coord.near(
            geo_coordinate2d(longitude(0.0009), latitude(0.001)), 1e-4))

    def test_operator_less_than(self):
        """Tests less than operator."""
        sw = geo_coordinate2d(longitude(-0.1), latitude(-0.1))
        ne = geo_coordinate2d(longitude(+0.1), latitude(+0.1))

        self.assertTrue(sw.get_longitude() < ne.get_longitude())
        self.assertTrue(sw.get_latitude() < ne.get_latitude())

    def test_operator_less_than_or_equal(self):
        """Tests less than or equal operator."""
        sw = geo_coordinate2d(longitude(-0.1), latitude(-0.1))
        ne = geo_coordinate2d(longitude(+0.1), latitude(+0.1))
        self.assertTrue(sw.get_longitude() <= ne.get_longitude())
        self.assertTrue(sw.get_latitude() <= ne.get_latitude())

        sw = geo_coordinate2d(longitude(-0.1), latitude(-0.1))
        ne = geo_coordinate2d(longitude(-0.1), latitude(-0.1))
        self.assertTrue(sw.get_longitude() <= ne.get_longitude())
        self.assertTrue(sw.get_latitude() <= ne.get_latitude())

    def test_operator_equal(self):
        """Tests for equality of geo_coordinate2d."""
        sw = geo_coordinate2d(longitude(-0.1), latitude(-0.1))
        ne = geo_coordinate2d(longitude(-0.1), latitude(-0.1))
        self.assertEquals(sw.get_longitude(), ne.get_longitude())
        self.assertEquals(sw.get_latitude(), ne.get_latitude())

File: test/test_geo_coordinate2d.cpp
#include <gtest/gtest.h>

#include "quad_keys/geo_coordinate2d.h"

#include <iostream>

namespace quad_keys
{
  class TestGeoCoordinate2d : public ::testing::Test
  {
  protected:
    virtual void SetUp() {}
    virtual void TearDown() {}
  };

  TEST(GEOCOORDINATE2D, default_ctor)
  {
    GeoCoordinate2d empty;
    EXPECT_EQ(Longitude(0.0), empty.get_longitude());
    EXPECT_EQ(Latitude(0.0), empty.get_latitude());
  }

  TEST(GEOCOORDINATE2D, ctor)
  {
    {
      GeoCoordinate2d coord(Longitude(0.0), Latitude(0.0));
      EXPECT_EQ(Longitude(0.0), coord.get_longitude());
      EXPECT_EQ(Latitude(0.0), coord.get_latitude());
    }

    {
      GeoCoordinate2d coord(Longitude(-180.001), Latitude(-95.0));
      EXPECT_EQ(Longitude(-180.0), coord.get_longitude());
      EXPECT_EQ(Latitude(-90.0), coord.get_latitude());
    }

    {
      GeoCoordinate2d coord(Longitude(-180.001), Latitude(95.0));
      EXPECT_EQ(Longitude(-180.0), coord.get_longitude());
      EXPECT_EQ(Latitude(90.0), coord.get_latitude());
    }

    {
      GeoCoordinate2d coord(Longitude(180.001), Latitude(95.0));
      EXPECT_EQ(Longitude(180.0), coord.get_longitude());
      EXPECT_EQ(Latitude(90.0), coord.get_latitude());
    }

    {
      GeoCoordinate2d coord(Longitude(180.001), Latitude(-95.0));
      EXPECT_EQ(Longitude(180.0), coord.get_longitude());
      EXPECT_EQ(Latitude(-90.0), coord.get_latitude());
    }
  }

  TEST(GEOCOORDINATE2D, near)
  {
    {
      Longitude lon(0.001);
      EXPECT_TRUE(lon.near(Longitude(0.0009), 1e-3));
    }

    {
      Longitude lon(0.001);
      EXPECT_FALSE(lon.near(Longitude(0.0009), 1e-4));
    }

    {
      Latitude lat(0.001);
      EXPECT_TRUE(lat.near(Latitude(0.0009), 1e-3));
    }

    {
      Latitude lat(0.001);
      EXPECT_FALSE(lat.near(Latitude(0.0009), 1e-4));
    }

    {
      GeoCoordinate2d coord(Longitude(0.001), Latitude(0.001));
      EXPECT_TRUE(coord.near(GeoCoordinate2d(Longitude(0.0009),
        Latitude(0.0009)), 1e-3));
    }

    {
      GeoCoordinate2d coord(Longitude(0.001), Latitude(0.001));
      EXPECT_FALSE(coord.near(GeoCoordinate2d(Longitude(0.001),
        Latitude(0.0009)), 1e-4));
    }

    {
      GeoCoordinate2d coord(Longitude(0.001), Latitude(0.001));
      EXPECT_FALSE(coord.near(GeoCoordinate2d(Longitude(0.0009),
        Latitude(0.001)), 1e-4));
    }
  }

  TEST(GEOCOORDINATE2D, operator_lt)
  {
    GeoCoordinate2d sw(Longitude(-0.1), Latitude(-0.1));
    GeoCoordinate2d ne(Longitude(+0.1), Latitude(+0.1));

    EXPECT_TRUE(sw.get_longitude() < ne.get_longitude());
    EXPECT_TRUE(sw.get_latitude() < ne.get_latitude());
  }

  TEST(GEOCOORDINATE2D, operator_lteq)
  {
    {
      GeoCoordinate2d sw(Longitude(-0.1), Latitude(-0.1));
      GeoCoordinate2d ne(Longitude(+0.1), Latitude(+0.1));

      EXPECT_TRUE(sw.get_longitude() <= ne.get_longitude());
      EXPECT_TRUE(sw.get_latitude() <= ne.get_latitude());
    }

    {
      GeoCoordinate2d sw(Longitude(-0.1), Latitude(-0.1));
      GeoCoordinate2d ne(Longitude(-0.1), Latitude(-0.1));

      EXPECT_TRUE(sw.get_longitude() <= ne.get_longitude());
      EXPECT_TRUE(sw.get_latitude() <= ne.get_latitude());
    }
  }

  TEST(GEOCOORDINATE2D, operator_eq)
  {
    GeoCoordinate2d sw(Longitude(-0.1), Latitude(-0.1));
    GeoCoordinate2d ne(Longitude(-0.1), Latitude(-0.1));

    EXPECT_TRUE(sw.get_longitude() == ne.get_longitude());
    EXPECT_TRUE(sw.get_latitude() == ne.get_latitude());
  }
}
